subprojects/internal-testing/src/main/groovy/org/gradle/integtests/fixtures/FlakyTestInterceptor.groovy

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.fixtures
+
+import org.gradle.util.FlakyTest
+
+import java.lang.reflect.Method
+import java.util.function.Consumer
+
+/**
+ * A flaky test interceptor to be reused by both JUnit4 {@link FlakyTestRule} and Spock {@link FlakyTestExtension}.
+ * By default, in "normal" mode, it skips all tests marked by `@FlakyTest`; in "flakyTestQuarantine" mode, it skip
+ * all tests except those marked by `@FlakyTest`.
+ */
+class FlakyTestInterceptor {
+    private static final boolean FLAKY_TEST_QUARANTINE_MODE = Boolean.getBoolean("org.gradle.flakyTestQuarantine")
+    private final Method testMethod
+    /**
+     * The action performed to proceed the test.
+     */
+    private final Runnable proceedAction
+    /**
+     * The action performed to skip the test, with an optional parameter: @FlakyTest.issue
+     */
+    private final Consumer<String> skipAction
+
+    FlakyTestInterceptor(Method testMethod, Runnable proceedAction, Consumer<String> skipAction) {
+        this.testMethod = testMethod
+        this.proceedAction = proceedAction
+        this.skipAction = skipAction
+    }
+
+    void intercept() {
+        FlakyTest flakyTestAnnotation = findAnnotation(testMethod)
+        String issue = flakyTestAnnotation?.issue()
+        boolean isFlaky = flakyTestAnnotation != null && flakyTestAnnotation.flakyIf().newInstance(null, null).call()
+        if (FLAKY_TEST_QUARANTINE_MODE) {
+            if (isFlaky) {
+                proceedAction.run()
+            } else {
+                skipAction.accept(issue ?: "Skip because of flakiness")
```

Comment:
```suggestion
                skipAction.accept(issue ?: "Skipped because the test is non-flaky")
```

For flaky test quarantine, we skip the non-flaky tests.